main에 devlop 브랜치 머지

rwaming opened the pull request.

rwaming committed:
check commit practice

rwaming committed:
start use github

rwaming committed:
please

rwaming committed:
start to use GitHub

rwaming committed:
check commit from another PC

rwaming committed:
231211 commit

rwaming committed:
231212 start

rwaming committed:
231212 edti - drag To Do(no order change)

rwaming committed:
231212 order change as btn moves

rwaming committed:
231212 done edit mode - change order

rwaming committed:
edit clicked, multiply eventListener 'window.addEventListener('mouseup',drag_finish)'

catn-ip commented on css/ToDo.css:
```diff
@@ -0,0 +1,179 @@
+    display: block;
+    font: normal 14px/1.5em 'Noto Sans KR',sans-serif;
+    text-align: center;
+}
```
전역 css 설정을 할때는 기본값으로 설정할때만 사용하며, 최소한으로 쓰는것을 추천합니다.
또한, body속성 또는 html 속성에 기본 폰트를 설정하면 나머지 요소들도 기본적으로 body의 속성을 상속받아오므로
아래와 같이 변경하는 것을 추천합니다.
```suggestion
body{ /* 텍스트 */
    font: normal 14px/1.5em 'Noto Sans KR',sans-serif;
}
```

catn-ip commented on js/ToDo.js:
```diff
@@ -0,0 +1,305 @@
+window.addEventListener('load',function(){
```
[링크](https://ko.javascript.info/script-async-defer)를 읽어보고 수정하길 바람

catn-ip commented on js/ToDo.js:
```diff
@@ -0,0 +1,305 @@
+    // typeBox_Todo value
+        function typeBox_TD_txt() { // 유저가 input에 입력한 ToDo
+            return document.querySelector('#typeBox_Todo').value;
+        }
```
document.querySelector('#typeBox_Todo')를 변수로 사용하고 있으므로 아래와 같이 수정한다면 보다 짧게 코드를 작성하며 훨씬 더 쉽게 이해할수 있도록 만들수 있음
```suggestion
        var typeBox_TD = document.querySelector('#typeBox_Todo');

    // typeBox_Todo value
        function typeBox_TD_txt() { // 유저가 input에 입력한 ToDo
            return typeBox_TD.value;
        }
```

catn-ip commented on js/ToDo.js:
```diff
@@ -0,0 +1,305 @@
+        }
+        function TD_txt_blank() { // 빈칸인지 체크
+            return typeBox_TD_txt().replace(/\s*/g,'')=='';
+        }
```
[링크](https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Global_Objects/String/trim)를 보고 적용시키면 더욱 좋을것 같음

catn-ip commented on js/ToDo.js:
```diff
@@ -0,0 +1,305 @@
+            new_btn_edit.innerHTML = 'edit';
+            new_btn_edit.addEventListener('click',open_edit) // edit 클릭시
+            return new_btn_edit;
+        }
```
edit 버튼을 html에 넣어둔 후 display: none과 같은 속성으로 해당 html을 보이지 않도록 설정하는 방식이 보다 편리해보입니다.
추가로 display: none과 visibility: hidden의 차이를 확인해보면 더욱더 좋을것 같습니다.

catn-ip commented on js/ToDo.js:
```diff
@@ -0,0 +1,305 @@
+        // btn_edit 인식
+            function btn_edit(){
+                return document.querySelector('.btn_edit');
+            }
```
값이 존재하지 않을때를 걱정하고 있는것이라면
querySelector에서 값이 없을때 반환해주는 null값을 관리할수 있도록
옵셔널체이닝이라는것을 공부해보는것을 추천합니다~
이후에 함수를 호출하는 코드를 상황에 맞추어 변수 또는 직접 document.querySelector()로 변경해주세요

rwaming replied:
옵셔널체이닝은 곧 공부하겠습니당

catn-ip commented on js/ToDo.js:
```diff
@@ -0,0 +1,305 @@
+                            list_style.style.opacity = '1';
+                            list_style.style.cursor = 'pointer';
+                            list_style.style.userSelect = 'none';
+                        }, 300);
```
우선적으로 transition 또는 animation을 css에 선언하는 방식으로 수정해보고
원하는 형태로 만들기 너무 어려우면 다시한번 질문해주시면 감사하겠습니다.

RWAMing replied:
진행중입니당

rwaming replied:
애니메이션은 맘대로 안돼서 classList.add랑 remove로 해결했고, 다만 opacity도 CSS를 이용하면 원하는 트랜지션 느낌을 얻을 수 없어서, opacity만 list_style.style.opacity = '웅앵';으로 사용했습니다.

catn-ip commented on js/ToDo.js:
```diff
@@ -0,0 +1,305 @@
+                    lists_style_edit.forEach(drag_ready); // 모든 To Do List 드래그 준비
+                }
+            // 드래그 이벤트리스너 준비
+                function drag_ready(element_drag){
+
+                    if(mode_edit){
+                        element_drag.addEventListener('mousedown',drag_start);
+                    }else{
+                        element_drag.removeEventListener('mousedown',drag_start);
+                        return;
+                    }
```
drag_start가 drag_ready 함수가 실행될 때 마다 새롭게 선언되고 있습니다.
이에 따라 removeEventListener가 정상적으로 작동될 수 없습니다.

catn-ip commented on js/ToDo.js:
```diff
@@ -0,0 +1,305 @@
+        '좋아하는 소비를 위해 저축하기', // 8
+        '운동하러 나가기' // 9
+    ];
+    function random_10() { // 랜덤 index번호 (0~9)
+        var i = Math.floor(Math.random()*10);
+        if(i==10){ // 낮은 확률인 10 제거
+            return 9;
+        }else{return i;}
+    }
```
불필요한 코드가 실행중입니다.
```suggestion
    function random_10() { // 랜덤 index번호 (0~9)
        return Math.floor(Math.random()*9);
    }
```

rwaming replied:
Math.random()이 0에서 1중에 랜덤숫자를 꺼낸다고 하는데, 이 상황에서 *10에다가 floor를 사용하면, 0~9가 될 확률은 같고 10이 될 확률은 너무 낮아서, 아예 10이 나왔을 경우에는 없던 일로 쳐버리려고 저렇게 했는데 별루일까요?

catn-ip replied:
0 ~ 9까지 나올 확률은 모두 비슷합니다.
아래 코드를 통해 실험해보면 비슷한 확률로 0~9까지가 나오는것을 알 수 있습니다.

```javascript
function calcRandomPercent() {
    return Array.from({length: 1000000}, 
                      () => Math.floor(Math.random() * 9))
          .reduce((acc, cur) => {
              acc[cur]++;
              return acc;
          }, [0,0,0,0,0,0,0,0,0])
}

calcRandomPercent().map(v => v / 1000000)
```

rwaming replied:
무슨 코드인지 모르겠어서 사실 아직 이해를 못했지만 *9를 하면 결론은 0~9 확률이 비슷하다고 이해하겠습니다.
해당 코드 공부해서 더 잘 이해하고 기억해보도록 하겠습니다 👍

catn-ip reviewed (commented):
var, let, const의 차이를 확인하고 적절하게 변경해보기를 추천합니다.

style을 javascript 코드에서 변경하지 않고 css에 class를 사용하여 선언 후 class를 변경하는 방식을 사용해보는것을 추천합니다.
(추후 유지보수에 보다 유리하기 때문)

단순히 다른 함수의 값을 반환하는 함수는 전부 반환하는 함수를 사용할 수 있도록 코드를 수정하는것을 추천합니다.
만약 반환 값이 변경되지 않는 경우 변수로 변경하여 재사용 하는 방식으로 변경하는것을 추천합니다.

rwaming committed:
js에서 css로 옮기는 작업

RWAMing commented:
옵셔널체이닝 공부랑 function random_10 제외하고 모두 완료했습니당

rwaming committed:
Fix: removeEventListener, var 수정

1. 함수 내 함수를 밖으로 꺼냄.
2. 외부에서 변수 선언. 함수 내부에서 그 변수에 값을 대입해서 다른 함수에서도 사용하도록 함.
3. 모든 var를 let과 const로 변경

Resolves: #12
Ref: #9, #10

rwaming committed:
Style: 여백과 주석 정리

html, js 여백정리 / CSS 주석 및 코드 순서 정리

.

rwaming committed:
style: console.log제거, html줄바꿈

제곧내

.

rwaming committed:
check용 commit
